WIP Feature/drop envelope database

NateWilliams2 opened the pull request:
- [x] Index nullifiers -> voteTxs on scrutinizer
- [ ] Hash envelopes on VoteTree
- [ ] Remove `state.Envelope()` (or rename it to `state.EnvelopeHash()`) and switch to `scrutinizer.GetEnvelope()`  instead
- [ ] Rename scrutinizer to indexer `find -type f -exec sed -i s/scrutinizer/indexer/g {}\;`

p4u committed:
vochaintest: add new flag --checkNulliriers

Checking all nullifiers one by one is very slow and costly.
We can just relay on the vote envelope height and the final
results in order to verify the votes.

The flag is false by default.

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
scrutinizer refactor and optimizations

This refactor is product of a deep benchmarking performed during some days
with a huge load of votes (up to 10 Million). Some weakness have been
found so they are addressed here.

Use type [][]*big.Int instead of protobuf bytes for storing the vote
results since it's more efficient when performing operations (adding
votes).

Instead of make the scrutinizer count always all live votes, count only
live votes after fastsync is finished. This increases the bootstrap time
for the Vochain.

Make the live votes counting async, so it does not have a huge impact on
the consensus block time.

Handle the computation of final results for unencrypted processes the
same way we compute the encrypted ones. The cost of counting all votes
at once is much lower than computing one-by-one as they are introduced in the
blockchain. So when final results computation is executed, the live
votes result will be override and the goroutines that already need to
compute live votes will be halted.

When counting live votes do not make 1 db transaction per vote but
aggregate them by processId and make a single transaction per block and
processId.

Show the last votes per minute (vxm) on the [vochain info] log message.

Signed-off-by: p4u <pau@dabax.net>

vdo committed:
Add compose post logs when fail

p4u committed:
apply review comments

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
scrutinizer: finish the recovery mechanism

Since live votes are not computed during Fast Sync and we don't know
which envelopes have been counted when starting the blockchain, we need
to execute a bootstrap live results recovery.

Apply also some review comments.

Signed-off-by: p4u <pau@dabax.net>

mvdan committed:
don't re-initialize the logger in each test

The logger is global, and already initialized by default as of
6d68b9cf96.

These tests were re-initializing the logger while others were still
logging stuff, causing data races:

	WARNING: DATA RACE
	Read at 0x0000024de2c0 by goroutine 81:
	  go.vocdoni.io/dvote/log.Infof()
	      /src/dvote/log/log.go:157 +0xdb
	  go.vocdoni.io/dvote/vochain/scrutinizer.(*Scrutinizer).afterSyncBootstrap()
	      /src/dvote/vochain/scrutinizer/scrutinizer.go:98 +0xcb

	Previous write at 0x0000024de2c0 by goroutine 261:
	  go.vocdoni.io/dvote/log.Init()
	      /src/dvote/log/log.go:41 +0x447
	  go.vocdoni.io/dvote/vochain/scrutinizer.TestResults()
	      /src/dvote/vochain/scrutinizer/scrutinizer_test.go:219 +0x72
	  testing.tRunner()
	      /tip/src/testing/testing.go:1235 +0x22d

p4u committed:
disable Tendermint tx indexer and fix scrutinizer race

Since we don't use the tendermint transaction indexer it is worth to
disable it. Its impact on performance and disk size is very important.

On scrutinizer do not execute AfetSyncBootstrap on the constructor, so
it will only be enabled if the developer wants. For testing its not
required and its the cause for some data races.

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
vochain state: add txBlockCounter

The txBlockCounter can be used by the event callbacks to perform
operations over specific transactions that can be fetch from the
tendermint blockStore.

Signed-off-by: p4u <pau@dabax.net>

NateWilliams2 committed:
Add txIndex to transaction callbacks

NateWilliams2 committed:
Add txIndex to votePool

NateWilliams2 committed:
Use scrutinizer/blockStore to fetch envelopes from router

p4u commented on router/voteCallbacks.go:
```diff
@@ -106,21 +106,31 @@ func (r *Router) getEnvelopeStatus(request routerRequest) {
 		r.sendError(request, fmt.Sprintf("cannot get envelope: (%s)", err))
 		return
 	}
+	block := r.vocapp.Node.BlockStore().LoadBlock(txRef.BlockHeight)
```
I think the `State` should have this methods abstracted.
+ state.GetBlockByHeight(height)
+ state.GetBlockByHash(hash)
+ state.GetTx(height, txIndex)

I'm not sure the Tendermint  BlockStore is safe, if not we should add a `sync.Mutex`

The Scrutinizer then would have a method for geting a specific envelope (already Unmarshaled)

+ scrutinizer.GetEnvelope(nullifier)

NateWilliams2 replied:
It seems like the BlockStore is safe for this:
```go
type BlockStore struct {
	db dbm.DB

	// mtx guards access to the struct fields listed below it. We rely on the database to enforce
	// fine-grained concurrency control for its data, and thus this mutex does not apply to
	// database contents. The only reason for keeping these fields in the struct is that the data
	// can't efficiently be queried from the database since the key encoding we use is not
	// lexicographically ordered (see https://github.com/tendermint/tendermint/issues/4567).
	mtx    tmsync.RWMutex
	base   int64
	height int64
}
```

p4u commented on vochain/scrutinizer/db.go:
```diff
@@ -46,6 +46,14 @@ type Entity struct {
+type VoteReference struct {
+	Nullifier    types.HexBytes `badgerholdKey:"Nullifier"`
+	ProcessId    types.HexBytes `badgerholdIndex:"ProcessID"`
+	BlockHeight  int64
```
Just Height should be fine

p4u commented on vochain/scrutinizer/vote.go:
```diff
@@ -27,6 +27,14 @@ var ErrNoResultsYet = fmt.Errorf("no results yet")
+func (s *Scrutinizer) GetVoteReference(nullifier []byte) (*VoteReference, error) {
+	txRef := &VoteReference{}
+	err := s.db.FindOne(txRef, badgerhold.Where(badgerhold.Key).Eq(nullifier))
+	return txRef, err
```
You can do: `return txRef, s.db.FindOne(txRef, badgerhold.Where(badgerhold.Key).Eq(nullifier))`

p4u commented on vochain/scrutinizer/vote.go:
```diff
@@ -174,6 +182,17 @@ func (s *Scrutinizer) addLiveVote(envelope *models.Vote, results *Results) error
+// addVoteIndex adds the nullifier reference to the kv for fetching vote Txs from BlockStore.
+// This method is triggered by Commit callback for each vote added to the blockchain.
+func (s *Scrutinizer) addVoteIndex(nullifier, pid []byte, blockHeight int64, txIndex int32) error {
+	_, err := s.VochainState.Process(pid, false)
```
There is no need for this call. We should in general try to minimize the number of database transactions.

p4u commented on vochain/scrutinizer/vote.go:
```diff
@@ -174,6 +182,17 @@ func (s *Scrutinizer) addLiveVote(envelope *models.Vote, results *Results) error
+		return fmt.Errorf("cannot get process %x: %w", pid, err)
+	}
+	voteRef := &VoteReference{Nullifier: nullifier, ProcessId: pid, BlockHeight: blockHeight, TxIndex: txIndex}
+	return s.db.Insert(nullifier, voteRef)
```
no need to declare voteRef on a separated line

p4u committed:
import the vochain BaseApplication on Scrutinizer

Instead of State, import the full BaseApplication, so Scrutinizer can
access to the tendermint BlockStore() in order to fetch full blocks and
transactions.

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
use scrutinizer/indexer for querying envelope data

Signed-off-by: p4u <pau@dabax.net>

NateWilliams2 committed:
Enable GetEnvelopeHeight for all processes

NateWilliams2 committed:
Rename BlockHeight -> Height, use VoteWithIndex struct in VotePool

p4u committed:
scrutinizer: split votePool and voteIndexPool

While votePool only count votes after FastSync, voteIndexPool needs
to index all votes (during fastSync too).

This fixes GetEnvelopeHeight()

Signed-off-by: p4u <pau@dabax.net>

NateWilliams2 committed:
Test envelope height, reference

p4u committed:
scrutinizer optimizations

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
scrutinizer: remove addVote()

AddVote() is now a method of the Results type.

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
add vote tx index database benchmark

Signed-off-by: p4u <pau@dabax.net>

jordipainan committed:
add support for rinkeby

jordipainan committed:
update token-storage-proof smartcontract bindigs

p4u committed:
fix benchmark and remove some indexes

Signed-off-by: p4u <pau@dabax.net>

NateWilliams2 committed:
App: implement block getters as configurable callbacks

NateWilliams2 committed:
Scrutinizer: fix some tests, save process opts in db

NateWilliams2 committed:
Scrutinizer: Fix results calculation

NateWilliams2 committed:
App: use SendTx callback method

NateWilliams2 committed:
Benchmark process indexing

NateWilliams2 committed:
Implement MockBlockStore sync.Map

jordipainan committed:
add support for envelopeType field costFromWeight:
- Updated contract bindings
- Modify extractEnvelopeType function accordingly
- Change maximum envelopeType value from 15 to 31 (0b00011111)

p4u committed:
update protobuf bindings

Signed-off-by: p4u <pau@dabax.net>
